sound/drivers/pimidi.c

```diff
@@ -0,0 +1,1050 @@
+			flush_delayed_work(&instance->midi_port[substream->number].out_handler);
+		printd(instance, "Flushed");
+	} while (!snd_rawmidi_transmit_empty(substream) && time_before(jiffies, deadline));
+
+	printd(instance, "Done!");
+}
+
+static int pimidi_midi_output_close(struct snd_rawmidi_substream *substream)
+{
+	struct pimidi_instance *instance = substream->rmidi->private_data;
+	struct pimidi_midi_port *midi_port = &instance->midi_port[substream->number];
+
+	mutex_lock(&midi_port->out_lock);
+	clear_bit(substream->stream, &midi_port->enabled_streams);
+	mutex_unlock(&midi_port->out_lock);
+	return 0;
+}
+
+static int pimidi_midi_input_close(struct snd_rawmidi_substream *substream)
+{
+	struct pimidi_instance *instance = substream->rmidi->private_data;
+	struct pimidi_midi_port *midi_port = &instance->midi_port[substream->number];
+
+	mutex_lock(&midi_port->in_lock);
+	clear_bit(substream->stream, &midi_port->enabled_streams);
+	mutex_unlock(&midi_port->in_lock);
+	return 0;
+}
+
+static void pimidi_midi_input_trigger(struct snd_rawmidi_substream *substream, int up)
+{
+	struct pimidi_instance *instance = substream->rmidi->private_data;
+
+	printd(instance, "(%d, %d, %d)", substream->stream, substream->number, up);
+
+	if (up == 0) {
+		clear_bit(substream->stream, &instance->midi_port[substream->number].enabled_streams);
+		cancel_work_sync(&instance->midi_port[substream->number].in_handler);
+	} else {
+		set_bit(substream->stream, &instance->midi_port[substream->number].enabled_streams);
+		if (!instance->stopping)
+			queue_work(pimidi_global.work_queue, &instance->midi_port[substream->number].in_handler);
+	}
+}
+
+static void pimidi_get_port_info(struct snd_rawmidi *rmidi, int number, struct snd_seq_port_info *seq_port_info)
+{
+	printd_g("%p, %d, %p", rmidi, number, seq_port_info);
+	seq_port_info->type =
+		SNDRV_SEQ_PORT_TYPE_MIDI_GENERIC |
+		SNDRV_SEQ_PORT_TYPE_HARDWARE |
+		SNDRV_SEQ_PORT_TYPE_PORT;
+	strscpy(seq_port_info->name, number == 0 ? "pimidi-a" : "pimidi-b", sizeof(seq_port_info->name));
+	seq_port_info->midi_voices = 0;
+}
+
+static const struct snd_rawmidi_global_ops pimidi_midi_ops = {
+	.get_port_info = pimidi_get_port_info,
+};
+
+static int pimidi_midi_open(struct snd_rawmidi_substream *substream)
+{
+	printd_g("(%p) stream=%d number=%d", substream, substream->stream, substream->number);
+	return 0;
+}
+
+static const struct snd_rawmidi_ops pimidi_midi_output_ops = {
+	.open = pimidi_midi_open,
+	.close = pimidi_midi_output_close,
+	.trigger = pimidi_midi_output_trigger,
+	.drain = pimidi_midi_output_drain,
+};
+
+static const struct snd_rawmidi_ops pimidi_midi_input_ops = {
+	.open = pimidi_midi_open,
+	.close = pimidi_midi_input_close,
+	.trigger = pimidi_midi_input_trigger,
+};
+
+static int pimidi_register(struct pimidi_instance *instance)
+{
+	int err = 0;
+
+	mutex_lock(&pimidi_global.lock);
+	printd(instance, "Registering...");
+	if (!pimidi_global.reset_gpio) {
```

Comment:
This looks very weird. You have unique reset gpios defined in the overlay for each of the address options, and yet you have this global state that after the first device has registered and acquired the GPIO, all subsequent ones don't bother. Why?

In fact why do you need this global state at all? I can't see any actually shared resource between multiple devices.

Reply:
Take it back - I see it's the `data-ready` GPIOs that are unique, and reset isn't. That's rather a shame as it makes the driver far more complex.